### ISSUE TITLE
[JENKINS-5977] Dead lock condition due to pipe clogging

### ISSUE THREAD

**Issue by timja-bot** [timja-org/jenkins-gh-issues-poc-11-04#580]
<p>The way the remoting is written, a certain race condition can result in a dead lock.</p>

<p>The root cause of this is that Channel.ReaderThread performs blocking write operation to OutputStream when it's executing the ProxyOutputStream.Chunk command. This synchronous behavior was necessary to ensure the proper in-order data arrival guarantee.</p>

<p>Now, the situation is, the master initiates a FilePath.copyRecursiveTo from the master to a slave. This creates a piped stream pair over Channel, where the sender on the master are directly sending data over Channel, which gets to FastPipedOutputStream on the slave, then the forwarded Callable on the slave is reading from the corresponding FastPipedInputStream.</p>

<p>The callable executing on the slave needs to load a class from the master, which results in this callable making an callback to the master to fetch a class file. But while this class file arrives as a Command, the master can send enough byte sequence to FastPipedOutputStream to the point that the buffer fills up and the write operation blocks.</p>

<p>At this point, the class file is stuck in the network waiting to be read by the command reader thread, but that won't be able to do that until there'll be some space in the buffer, and there won't be any space in the buffer until the class file is retrieved, hence the dead lock.</p>

---
<details><summary><i>Originally reported by <a title="Kohsuke Kawaguchi" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-5977" target="_blank">Dead lock condition due to pipe clogging</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Kohsuke Kawaguchi" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2010-10-05T14:26:02+00:00
<li><b>imported</b>: 2025-11-04
</ul></i>
</details>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>The reverse chronological sequence of the log showing what the master has sent. Looking at the slave stack trace, it hasn't read a single byte yet, but the master has sent in the 370K data in "Pipe.Chunk(6,371226)", and the class file image afterward, which is never read.</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7900)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7988)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7908)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7804)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7740)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7996)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7888)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7856)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7964)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7984)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Response[retVal=hudson.remoting.RemoteClassLoader$ClassFile@5f471136,exception=null]

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Received RPCRequest(2,fetch2)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7768)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,8192)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,7928)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send Pipe.Chunk(6,371226)

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Received hudson.remoting.Pipe$ConnectCommand@47d7e1fb

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send UserRequest:hudson.FilePath$FileCallableWrapper@40af3623

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Received Response[retVal=hudson.remoting.UserResponse@145e62e7,exception=null]

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Send UserRequest:hudson.EnvVars$GetEnvVars@652dc559

Mar 17, 2010 6:13:02 PM hudson.remoting.Channel
FINE: Received Response[retVal=hudson.remoting.UserResponse@762c3f0,exception=null]

Mar 17, 2010 6:13:01 PM hudson.remoting.Channel
FINE: Send UserRequest:hudson.EnvVars$GetEnvVars@490606c0

Mar 17, 2010 6:13:01 PM hudson.remoting.Channel
FINE: Received Response[retVal=hudson.remoting.UserResponse@3fd15873,exception=null]

Mar 17, 2010 6:13:01 PM hudson.remoting.Channel
FINE: Send UserRequest:hudson.FilePath$FileCallableWrapper@26c76ec2
</pre>
</div></div>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in hudson<br/>
User: : mindless<br/>
Path:<br/>
 trunk/hudson/plugins/copyartifact/src/main/java/hudson/plugins/copyartifact/CopyArtifact.java<br/>
 trunk/hudson/plugins/copyartifact/src/main/webapp/<a href="https://issues.jenkins.io/browse/JENKINS-5977" title="Dead lock condition due to pipe clogging" class="issue-link" data-issue-key="JENKINS-5977"><del>JENKINS-5977</del></a>/deleteme.tmp<br/>
<a href="http://jenkins-ci.org/commit/28825" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/28825</a><br/>
Log:<br/>
<span class="error">&#91;copyartifact&#93;</span> <span class="error">&#91;FIXED JENKINS-5934&#93;</span> Add workaround for <a href="https://issues.jenkins.io/browse/JENKINS-5977" title="Dead lock condition due to pipe clogging" class="issue-link" data-issue-key="JENKINS-5977"><del>JENKINS-5977</del></a><br/>
so slaves won't hang on copy of larger artifacts.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Another possibly related pipe clogging problem, this time the reader thread on the master is clogging:</p>

<p>This is the thread on a slave that's sending stream. This is FilePath.read()</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>"pool-1-thread-415" prio=6 tid=0x031df800 nid=0x604 runnable [0x02e6f000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(Unknown Source)
	at java.net.SocketOutputStream.write(Unknown Source)
	at java.io.BufferedOutputStream.flushBuffer(Unknown Source)
	at java.io.BufferedOutputStream.write(Unknown Source)
	- locked <0x22fe3c60> (a java.io.BufferedOutputStream)
	at java.io.ObjectOutputStream$BlockDataOutputStream.drain(Unknown Source)
	at java.io.ObjectOutputStream$BlockDataOutputStream.write(Unknown Source)
	at java.io.ObjectOutputStream.writeArray(Unknown Source)
	at java.io.ObjectOutputStream.writeObject0(Unknown Source)
	at java.io.ObjectOutputStream.defaultWriteFields(Unknown Source)
	at java.io.ObjectOutputStream.writeSerialData(Unknown Source)
	at java.io.ObjectOutputStream.writeOrdinaryObject(Unknown Source)
	at java.io.ObjectOutputStream.writeObject0(Unknown Source)
	at java.io.ObjectOutputStream.writeObject(Unknown Source)
	at hudson.remoting.Channel.send(Channel.java:417)
	- locked <0x22fda8d0> (a hudson.remoting.Channel)
	at hudson.remoting.ProxyOutputStream.write(ProxyOutputStream.java:112)
	- locked <0x2303de50> (a hudson.remoting.ProxyOutputStream)
	at hudson.remoting.ProxyOutputStream.write(ProxyOutputStream.java:96)
	at hudson.Util.copyStream(Util.java:406)
	at hudson.FilePath$24.call(FilePath.java:1202)
	at hudson.FilePath$24.call(FilePath.java:1197)
	at hudson.remoting.UserRequest.perform(UserRequest.java:114)
	at hudson.remoting.UserRequest.perform(UserRequest.java:48)
	at hudson.remoting.Request$2.run(Request.java:270)
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.util.concurrent.FutureTask$Sync.innerRun(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at hudson.remoting.Engine$1$1.run(Engine.java:58)
	at java.lang.Thread.run(Unknown Source)
</pre>
</div></div>

<p>This is the thread on the master hanging because of this:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>"Channel reader thread: hb-slave-trunk3" Id=53 TIMED_WAITING on [B@18c8d4d
	at java.lang.Object.wait(Native Method)
	-  waiting on [B@18c8d4d
	at hudson.remoting.FastPipedOutputStream.write(FastPipedOutputStream.java:147)
	at hudson.remoting.FastPipedOutputStream.write(FastPipedOutputStream.java:122)
	at hudson.remoting.ProxyOutputStream$Chunk.execute(ProxyOutputStream.java:164)
	at hudson.remoting.Channel$ReaderThread.run(Channel.java:867)
</pre>
</div></div>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Details for my previous comment.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Full log for my previous comment is at <a href="http://issues.jenkins-ci.org/secure/attachment/19251/log.zip" class="external-link" target="_blank" rel="nofollow noopener">http://issues.jenkins-ci.org/secure/attachment/19251/log.zip</a></p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=glundh">glundh</a>:</i>
<p>Proof of concept fix:</p>

<p><a href="http://hudson.361315.n4.nabble.com/Discussion-regarding-JENKINS-5977-Dead-lock-condition-due-to-pipe-clogging-And-more-td2322497.html" class="external-link" target="_blank" rel="nofollow noopener">http://hudson.361315.n4.nabble.com/Discussion-regarding-JENKINS-5977-Dead-lock-condition-due-to-pipe-clogging-And-more-td2322497.html</a></p>

<p>Please review.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=glundh">glundh</a>:</i>
<p>Proof-of-concept fix.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>I'm testing a patch similar to what's proposed by GLundh, except it does throttling to avoid unbounded growth of the queue. Plan to integrate it to the trunk soon, but I'm bit cautious as the remoting affects many things.</p>

<p>If there are other people willing to test the modification, I'll post the modified binary. </p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=glundh">glundh</a>:</i>
<p>Kohsuke: Thanks for your comment. I'd be happy to take it for a spin on our test-server.</p>

<p>And I would also really like to see the code (if that's possible). Would you mind branching your fix on the SVN or github?</p>

<p>Br<br/>
Gustaf - Sony Ericsson</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in hudson<br/>
User: : kohsuke<br/>
Path:<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/Capability.java<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/Channel.java<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/Pipe.java<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/PipeWindow.java<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/ProxyOutputStream.java<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/SynchronousExecutorService.java<br/>
 trunk/hudson/main/remoting/src/test/java/hudson/remoting/ChannelRunner.java<br/>
 trunk/hudson/main/remoting/src/test/java/hudson/remoting/PipeTest.java<br/>
 trunk/hudson/main/remoting/src/test/java/hudson/remoting/RmiTestBase.java<br/>
 trunk/hudson/main/remoting/src/test/java/hudson/remoting/WithRunner.java<br/>
<a href="http://jenkins-ci.org/commit/34838" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/34838</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-5977&#93;</span> implemented a windowing mechanism to throttle the pipe usage, and the pipe operation no longer interferes with the main channel reader thread.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dogfood">dogfood</a>:</i>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.hudson-labs.org/images/16x16/yellow.gif" style="border: 0px solid black" /></span> <a href="http://ci.hudson-labs.org/job/hudson_main_trunk/263/" class="external-link" target="_blank" rel="nofollow noopener">hudson_main_trunk #263</a><br/>
     <span class="error">&#91;FIXED JENKINS-5977&#93;</span> implemented a windowing mechanism to throttle the pipe usage, and the pipe operation no longer interferes with the main channel reader thread.</p>

<p>kohsuke : <br/>
Files : </p>
<ul>
	<li>/trunk/hudson/main/remoting/src/test/java/hudson/remoting/ChannelRunner.java</li>
	<li>/trunk/hudson/main/remoting/src/test/java/hudson/remoting/RmiTestBase.java</li>
	<li>/trunk/hudson/main/remoting/src/main/java/hudson/remoting/Channel.java</li>
	<li>/trunk/hudson/main/remoting/src/test/java/hudson/remoting/WithRunner.java</li>
	<li>/trunk/hudson/main/remoting/src/main/java/hudson/remoting/PipeWindow.java</li>
	<li>/trunk/hudson/main/remoting/src/main/java/hudson/remoting/Pipe.java</li>
	<li>/trunk/hudson/main/remoting/src/main/java/hudson/remoting/ProxyOutputStream.java</li>
	<li>/trunk/hudson/main/remoting/src/test/java/hudson/remoting/PipeTest.java</li>
	<li>/trunk/hudson/main/remoting/src/main/java/hudson/remoting/SynchronousExecutorService.java</li>
	<li>/trunk/hudson/main/remoting/src/main/java/hudson/remoting/Capability.java</li>
</ul>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=glundh">glundh</a>:</i>
<p>1.   34838_hang_on_archiving.txt (18 kB)<br/>
2.   34838_hang_on_emma_publisher.txt (11 kB)</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olamy">olamy</a>:</i>
<p>here a new jstack.<br/>
This issue made 1.378 unusable.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in hudson<br/>
User: : kohsuke<br/>
Path:<br/>
 branches/rc/remoting/src/main/java/hudson/remoting/Channel.java<br/>
 branches/rc/remoting/src/main/java/hudson/remoting/PipeWindow.java<br/>
 branches/rc/remoting/src/main/java/hudson/remoting/ProxyOutputStream.java<br/>
 branches/rc/remoting/src/test/java/hudson/remoting/PipeTest.java<br/>
 trunk/www/changelog.html<br/>
<a href="http://jenkins-ci.org/commit/35460" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/35460</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-5977 JENKINS-7572&#93;</span> the incorrect use of WeakHashMap resulted in two instances of PipeWindows for the same OID.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=esmalling">esmalling</a>:</i>
<p>I too am seeing the increase/decrease logs in 1.377.  Too bad too since I need 1.379 to fix <a href="https://issues.jenkins.io/browse/JENKINS-7546" title="Getting AbstractMethodError on all top-level (non job) pages after 1.377 upgrade" class="issue-link" data-issue-key="JENKINS-7546"><del>JENKINS-7546</del></a></p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=esmalling">esmalling</a>:</i>
<p>Unable to test <a href="https://issues.jenkins.io/browse/JENKINS-7546" title="Getting AbstractMethodError on all top-level (non job) pages after 1.377 upgrade" class="issue-link" data-issue-key="JENKINS-7546"><del>JENKINS-7546</del></a> due to this issue in 1.379</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=swpalmer">swpalmer</a>:</i>
<p>I'm seeing the following exceptions after a successful build with Hudson 1.379. They seem to be related to the problems with the pipes to the slave nodes.</p>

<p>FATAL: command execution failed<br/>
hudson.util.IOException2: Failed to join the process<br/>
	at hudson.Proc$RemoteProc.join(Proc.java:355)<br/>
	at hudson.Launcher$ProcStarter.join(Launcher.java:280)<br/>
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:82)<br/>
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:58)<br/>
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:19)<br/>
	at hudson.model.AbstractBuild$AbstractRunner.perform(AbstractBuild.java:601)<br/>
	at hudson.model.Build$RunnerImpl.build(Build.java:174)<br/>
	at hudson.model.Build$RunnerImpl.doRun(Build.java:138)<br/>
	at hudson.model.AbstractBuild$AbstractRunner.run(AbstractBuild.java:416)<br/>
	at hudson.model.Run.run(Run.java:1280)<br/>
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)<br/>
	at hudson.model.ResourceController.execute(ResourceController.java:88)<br/>
	at hudson.model.Executor.run(Executor.java:137)<br/>
Caused by: java.util.concurrent.ExecutionException: hudson.remoting.RequestAbortedException: java.net.SocketException: Connection reset<br/>
	at hudson.remoting.Request$1.get(Request.java:218)<br/>
	at hudson.remoting.Request$1.get(Request.java:172)<br/>
	at hudson.remoting.FutureAdapter.get(FutureAdapter.java:55)<br/>
	at hudson.Proc$RemoteProc.join(Proc.java:347)<br/>
	... 12 more<br/>
Caused by: hudson.remoting.RequestAbortedException: java.net.SocketException: Connection reset<br/>
	at hudson.remoting.Request.abort(Request.java:257)<br/>
	at hudson.remoting.Channel.terminate(Channel.java:681)<br/>
	at hudson.remoting.Channel$ReaderThread.run(Channel.java:973)<br/>
Caused by: java.net.SocketException: Connection reset<br/>
	at java.net.SocketInputStream.read(SocketInputStream.java:168)<br/>
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:218)<br/>
	at java.io.BufferedInputStream.read(BufferedInputStream.java:237)<br/>
	at java.io.ObjectInputStream$PeekInputStream.peek(ObjectInputStream.java:2249)<br/>
	at java.io.ObjectInputStream$BlockDataInputStream.peek(ObjectInputStream.java:2542)<br/>
	at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2552)<br/>
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1297)<br/>
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:351)<br/>
	at hudson.remoting.Channel$ReaderThread.run(Channel.java:948)<br/>
FATAL: Unable to delete script file C:\DOCUME~1\Hudson\LOCALS~1\Temp\hudson8908395142501843194.bat<br/>
hudson.util.IOException2: remote file operation failed: C:\DOCUME~1\Hudson\LOCALS~1\Temp\hudson8908395142501843194.bat at hudson.remoting.Channel@15509e1:DCM-IG-01<br/>
	at hudson.FilePath.act(FilePath.java:749)<br/>
	at hudson.FilePath.act(FilePath.java:735)<br/>
	at hudson.FilePath.delete(FilePath.java:990)<br/>
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:92)<br/>
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:58)<br/>
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:19)<br/>
	at hudson.model.AbstractBuild$AbstractRunner.perform(AbstractBuild.java:601)<br/>
	at hudson.model.Build$RunnerImpl.build(Build.java:174)<br/>
	at hudson.model.Build$RunnerImpl.doRun(Build.java:138)<br/>
	at hudson.model.AbstractBuild$AbstractRunner.run(AbstractBuild.java:416)<br/>
	at hudson.model.Run.run(Run.java:1280)<br/>
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)<br/>
	at hudson.model.ResourceController.execute(ResourceController.java:88)<br/>
	at hudson.model.Executor.run(Executor.java:137)<br/>
Caused by: hudson.remoting.ChannelClosedException: channel is already closed<br/>
	at hudson.remoting.Channel.send(Channel.java:467)<br/>
	at hudson.remoting.Request.call(Request.java:105)<br/>
	at hudson.remoting.Channel.call(Channel.java:630)<br/>
	at hudson.FilePath.act(FilePath.java:742)<br/>
	... 13 more<br/>
Caused by: java.net.SocketException: Connection reset<br/>
	at java.net.SocketInputStream.read(SocketInputStream.java:168)<br/>
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:218)<br/>
	at java.io.BufferedInputStream.read(BufferedInputStream.java:237)<br/>
	at java.io.ObjectInputStream$PeekInputStream.peek(ObjectInputStream.java:2249)<br/>
	at java.io.ObjectInputStream$BlockDataInputStream.peek(ObjectInputStream.java:2542)<br/>
	at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2552)<br/>
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1297)<br/>
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:351)<br/>
	at hudson.remoting.Channel$ReaderThread.run(Channel.java:948)<br/>
FATAL: channel is already closed<br/>
hudson.remoting.ChannelClosedException: channel is already closed<br/>
	at hudson.remoting.Channel.send(Channel.java:467)<br/>
	at hudson.remoting.Request.call(Request.java:105)<br/>
	at hudson.remoting.Channel.call(Channel.java:630)<br/>
	at hudson.Launcher$RemoteLauncher.kill(Launcher.java:744)<br/>
	at hudson.model.AbstractBuild$AbstractRunner.run(AbstractBuild.java:421)<br/>
	at hudson.model.Run.run(Run.java:1280)<br/>
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)<br/>
	at hudson.model.ResourceController.execute(ResourceController.java:88)<br/>
	at hudson.model.Executor.run(Executor.java:137)<br/>
Caused by: java.net.SocketException: Connection reset<br/>
	at java.net.SocketInputStream.read(SocketInputStream.java:168)<br/>
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:218)<br/>
	at java.io.BufferedInputStream.read(BufferedInputStream.java:237)<br/>
	at java.io.ObjectInputStream$PeekInputStream.peek(ObjectInputStream.java:2249)<br/>
	at java.io.ObjectInputStream$BlockDataInputStream.peek(ObjectInputStream.java:2542)<br/>
	at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2552)<br/>
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1297)<br/>
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:351)<br/>
	at hudson.remoting.Channel$ReaderThread.run(Channel.java:948)<br/>
Page generated: Oct 5, 2010 9:11:10 AMHudson ver. 1.379</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Logging issue is covered in <a href="https://issues.jenkins.io/browse/JENKINS-7662" title="endless pipe logging" class="issue-link" data-issue-key="JENKINS-7662"><del>JENKINS-7662</del></a> and is already fixed. The connection drop problem is not a deadlock problem, so please file a separate issue for that. Thanks!</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: alanharder<br/>
Path:<br/>
 src/main/java/hudson/plugins/copyartifact/FilePathCopyMethod.java<br/>
<a href="http://jenkins-ci.org/commit/copyartifact-plugin/8b2ceaaeb1b58f6ab075e5cb260691c36e25fca4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/copyartifact-plugin/8b2ceaaeb1b58f6ab075e5cb260691c36e25fca4</a><br/>
Log:<br/>
Remove workaround code for <a href="https://issues.jenkins.io/browse/JENKINS-5977" title="Dead lock condition due to pipe clogging" class="issue-link" data-issue-key="JENKINS-5977"><del>JENKINS-5977</del></a> now that this issue is fixed<br/>
in the minimum core version required by this plugin.</p>
